lib/TickLib.sol

```diff
@@ -62,110 +63,109 @@ library LeafLib {
   // not for onchain use
   function eraseToTick(Leaf leaf, Tick tick) internal pure returns (Leaf) {
-    uint mask = ONES >> ((tick.posInLeaf() + 1) * OFFER_BITS * 2);
-    return Leaf.wrap(Leaf.unwrap(leaf) & mask);
+    unchecked {
+      uint mask = ONES >> ((tick.posInLeaf() + 1) * OFFER_BITS * 2);
+      return Leaf.wrap(Leaf.unwrap(leaf) & mask);
+    }
   }
 
   function eraseFromTick(Leaf leaf, Tick tick) internal pure returns (Leaf) {
-    uint mask = ~(ONES >> (tick.posInLeaf() * OFFER_BITS * 2));
-    return Leaf.wrap(Leaf.unwrap(leaf) & mask);
+    unchecked {
+      uint mask = ~(ONES >> (tick.posInLeaf() * OFFER_BITS * 2));
+      return Leaf.wrap(Leaf.unwrap(leaf) & mask);
+    }
   }
 
   function firstOfTick(Leaf leaf, Tick tick) internal pure returns (uint) {
-    return firstOfIndex(leaf, TickLib.posInLeaf(tick));
+    unchecked {
+      return firstOfIndex(leaf, TickLib.posInLeaf(tick));
+    }
   }
 
   function lastOfTick(Leaf leaf, Tick tick) internal pure returns (uint) {
-    return lastOfIndex(leaf, TickLib.posInLeaf(tick));
+    unchecked {
+      return lastOfIndex(leaf, TickLib.posInLeaf(tick));
+    }
   }
 
   function firstOfIndex(Leaf leaf, uint index) internal pure returns (uint) {
-    uint raw = Leaf.unwrap(leaf);
-    return uint(raw << (index * OFFER_BITS * 2) >> (256 - OFFER_BITS));
+    unchecked {
+      uint raw = Leaf.unwrap(leaf);
+      return uint(raw << (index * OFFER_BITS * 2) >> (256 - OFFER_BITS));
+    }
   }
 
   function lastOfIndex(Leaf leaf, uint index) internal pure returns (uint) {
-    uint raw = Leaf.unwrap(leaf);
-    return uint(raw << (OFFER_BITS * ((index * 2) + 1)) >> (256 - OFFER_BITS));
+    unchecked {
+      uint raw = Leaf.unwrap(leaf);
+      return uint(raw << (OFFER_BITS * ((index * 2) + 1)) >> (256 - OFFER_BITS));
+    }
   }
 
-  // TODO optimize with a hashmap, or:
-  // if > half: +=1; if += quarter: +=1, or:
-  // or nested if dichotomy
+  // Will check for the first position (0,1,2 or 3) that has a nonzero first-of-tick or a nonzero last-of-tick offer. Leafs where only one of those is nonzero are invalid anyway.
   function firstOfferPosition(Leaf leaf) internal pure returns (uint ret) {
-    uint offerId = Leaf.unwrap(leaf) >> (OFFER_BITS * 7);
-    if (offerId != 0) {
-      ret = 0;
-    } else {
-      offerId = Leaf.unwrap(leaf) << (OFFER_BITS * 2) >> (OFFER_BITS * 7);
-      if (offerId != 0) {
-        ret = 1;
-      } else {
-        offerId = Leaf.unwrap(leaf) << (OFFER_BITS * 4) >> (OFFER_BITS * 7);
-        if (offerId != 0) {
-          ret = 2;
-        } else {
-          offerId = Leaf.unwrap(leaf) << (OFFER_BITS * 6) >> (OFFER_BITS * 7);
-          if (offerId != 0) {
-            ret = 3;
-          }
-        }
+    unchecked {
+      assembly("memory-safe") {
+        ret := shl(1,gt(0xffffffffffffffffffffffffffffffff,leaf))
+        ret := or(ret,gt(0xffffffffffffffff,shr(shl(7,iszero(ret)),leaf)))
```

Comment:
I have not checked this. Let me know if I should

Reply:
see #544